### ISSUE TITLE
Event Messaging System

### ISSUE THREAD

**Issue by sible777** [sible777/SGEngine#3]
### EMS

After heavy thinking, I arrived at the conclusion that most of the functionality one may want to implement in a game with this engine would be vastly improved by using an event system.

This will allow for objects to be notified of when they should do something rather than just polling things constantly. 

#### Example:
The game world is one persistent "level" polling every game world object for updates would be cpu intensive and wasteful as well as inflexible. However with an EMS in place each event could "subscribe" so to speak to the sub systems it needs and those systems can just listen for events telling them what to do. So, in theory only the objects that need to be updated would be updated.

And EMS system would also allow for greater flexibility and separation from the engine when it comes to scripting game events such as doors opening, triggering cut scenes and so on.

**Comment by sible777** [comment]
@RobDanHill @Shatterwaltz @Interrollipsis 
Mentioning you guys for the sake of discussion of the topic at hand.

**Comment by Shatterwaltz** [comment]
Sounds like events are the way to go, checking every single object in the level for updates would be pretty slow.

**Comment by sible777** [comment]
Yeah I agree. It's also a solution to simplifying the code base, especially since there are several sub systems planned out. Such as: AudioManager, Physics, AssetsManager, UIManager, etc. 

It will be a good way to orchestrate the interactions between the systems.

I haven't quite worked out the best way to implement all of this. In particular the way I want to use the event system to script gameplay separately from the engine itself.

#### Concept - pseudo code
```
class Event
    typeStr - A string to identify the type of event this is
    eventOwner - A reference to the object that emitted or owns this event
    callback - A function that is registered to this event for execution after the event is handled.

    Event(typeStr, eventOwner, callback)
end class Event

class EventManagementSystem
    // Not yet designed
end class EMS
```

The EMS class will then be responsible for handling these events and passing the necessary data along to the right submodules. This means that no game object has to "know" about any part of the engine other than the event system. So, in theory they can really easily communicate with any part of the engine and it's up to each part of the engine to know how to handle these events.

Also, the reference to the object that owns the event is so that a submodule can modify it appropriately. E.g the collision module should be able to set an objects coordinates for collision resolution.
